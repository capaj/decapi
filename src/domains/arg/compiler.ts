import {
  GraphQLFieldConfigArgumentMap,
  GraphQLInputType,
  isInputType
} from 'graphql'

import { IArgsIndex, argRegistry } from './registry'

import 'reflect-metadata'

import { injectorRegistry } from '../inject/Inject'

import { Constructor, reflect } from 'typescript-rtti'
import { inferTypeFromRtti } from '../../services/utils/gql/types/inferTypeByTarget'
import { resolveType } from '../../services/utils/gql/types/typeResolvers'

export interface ICompileArgContextType {
  target: Constructor<Function>
  fieldName: string
  argumentTypes: Array<GraphQLInputType | null>
  registeredArgs: IArgsIndex
  onlyDecoratedArgs: boolean
}

function convertArgsArrayToArgsMap({
  target,
  fieldName,
  argumentTypes,
  registeredArgs = [],
  onlyDecoratedArgs
}: ICompileArgContextType): GraphQLFieldConfigArgumentMap {
  const fieldDescriptor = Object.getOwnPropertyDescriptor(
    target.prototype,
    fieldName
  )

  // in case of getters, field arguments are not relevant
  if (!fieldDescriptor || fieldDescriptor.get) {
    return {}
  }

  const argNames = reflect(target).getOwnMethod(fieldName).parameterNames

  if (!argNames || !argNames.length) {
    return {}
  }

  const argsMap: GraphQLFieldConfigArgumentMap = {}

  argNames.forEach((argName: string, index: number) => {
    const registeredArgConfig = registeredArgs[index]

    if (onlyDecoratedArgs && !registeredArgConfig) {
      return
    }
    const argConfig = registeredArgConfig || {}

    if (argConfig.name) {
      argName = argConfig.name
    }
    const argType = argumentTypes[index]
    if (!argType) {
      return
    }
    // don't publish args marked as auto Injected
    if (injectorRegistry.has(target, fieldName, index)) {
      return
    }

    argsMap[argName] = {
      type: argType,
      description: argConfig.description
    }
  })

  return argsMap
}

export function compileFieldArgs(
  target: Constructor<Function>,
  fieldName: string,
  onlyDecoratedArgs: boolean
): GraphQLFieldConfigArgumentMap | null {

  const getter = Object.getOwnPropertyDescriptor(
    target.prototype,
    fieldName
  )?.get

  if (getter) {
    return null
  }

  const registeredArgs = argRegistry.getAll(target)[fieldName]

  const injectedArgs = injectorRegistry.getAll(target)[fieldName]
  const reflected = reflect(target)
  const methodRtti = reflected.getOwnMethod(fieldName)
  if (!methodRtti) {
    return null
  }
  const args = methodRtti.parameterTypes

  const argumentTypes: Array<GraphQLInputType | null> = []

  for (let index = 0; index < args.length; index++) {
    const registeredArgConfig = registeredArgs && registeredArgs[index]
    if (injectedArgs && injectedArgs[index]) {
      argumentTypes[index] = null
    } else if (registeredArgConfig?.type) {
      // @ts-expect-error
      argumentTypes[index] = resolveType(registeredArgConfig.type, true, true)
    } else if (onlyDecoratedArgs && !registeredArgConfig) {
      argumentTypes[index] = null
    } else {
      const rtti = args[index]
      const inferredType = inferTypeFromRtti(rtti)
      // @ts-expect-error
      argumentTypes[index] = inferredType
      argRegistry.set(target, [fieldName, index], {
        ...registeredArgConfig,
        inferredType,
        argIndex: index
      })
    }
  }

  const compileFieldArgContext = {
    target,
    fieldName,
    argumentTypes,
    registeredArgs,
    onlyDecoratedArgs
  }

  return convertArgsArrayToArgsMap(compileFieldArgContext)
}
