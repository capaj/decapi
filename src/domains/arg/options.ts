export interface IArgOptions {
  description?: string
  type?: any
  nullable?: boolean
  name?: string
}

export const defaultArgOptions: IArgOptions = {
  nullable: false
}
