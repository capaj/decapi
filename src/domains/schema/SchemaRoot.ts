export {
  schemaRootsRegistry,
  mutationFieldsRegistry,
  queryFieldsRegistry
} from './registry.js'
import { schemaRootsRegistry } from './registry.js'

export { compileSchema } from './compiler.js'
export { isSchemaRoot, getSchemaRootInstance } from './services.js'

export function SchemaRoot(config: object = {}): ClassDecorator {
  return (target) => {
    schemaRootsRegistry.set(target, config)
  }
}
