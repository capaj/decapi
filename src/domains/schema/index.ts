export {
  schemaRootsRegistry,
  mutationFieldsRegistry,
  queryFieldsRegistry,
} from './registry'
import { schemaRootsRegistry, SchemaRootConfig } from './registry'

// import { compileSchema } from './compiler';
export { compileSchema } from './compiler'
export { Query, Mutation } from './rootFields'
export { isSchemaRoot, getSchemaRootInstance } from './services'

export function SchemaRoot(config: SchemaRootConfig = {}): ClassDecorator {
  return (target) => {
    schemaRootsRegistry.set(target, config)
  }
}
