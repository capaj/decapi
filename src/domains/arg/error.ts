import { reflect } from 'typescript-rtti'
import { BaseError } from '../../services/error.js'

import { ITargetAndField } from './compiler.js'
// TODO use again when typescript rtti is fixed
export class ArgError extends BaseError {
  constructor(msg: string, ctx: ITargetAndField, argIndex: number) {
    const { target, fieldName } = ctx
    const paramNames = reflect(target).getOwnMethod(fieldName).parameterNames
    const paramName = paramNames[argIndex]
    const fullMsg = `@Type ${target.name}.${fieldName}(${paramName} <-------): ${msg}`
    super(fullMsg)
    this.message = fullMsg
  }
}
