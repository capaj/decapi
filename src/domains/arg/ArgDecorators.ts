import { argRegistry } from './registry'
export { compileFieldArgs } from './compiler'
import { IArgOptions, defaultArgOptions } from './options'

export function Arg(options: IArgOptions = {}): ParameterDecorator {
  return (target: Object, fieldName: string, argIndex: number) => {
    argRegistry.set(target.constructor, [fieldName, argIndex], {
      ...defaultArgOptions,
      ...options,
      argIndex
    })
  }
}

/**
 * a shorthand for @Arg({isNullable: true})
 */
export function ArgNullable(options: IArgOptions = {}): ParameterDecorator {
  return Arg({ ...options, isNullable: true })
}
