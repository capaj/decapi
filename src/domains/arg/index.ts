import { argRegistry } from './registry'
export { compileFieldArgs } from './compiler'
import { ArgOptions, defaultArgOptions } from './options'

export function Arg(options: ArgOptions = {}): ParameterDecorator {
  return (target: Object, fieldName: string, argIndex: number) => {
    argRegistry.set(target.constructor, [fieldName, argIndex], {
      ...defaultArgOptions,
      ...options,
      argIndex,
    })
  }
}
