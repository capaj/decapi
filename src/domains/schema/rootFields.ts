import {
  queryFieldsRegistry,
  mutationFieldsRegistry,
  schemaRootsRegistry
} from './registry'
import { SchemaFieldError } from './error'
import { compileFieldConfig, IFieldOptions, Field } from '../field/Field'

function validateRootSchemaField(targetInstance: Object, fieldName: string) {
  if (
    !(targetInstance as any)[fieldName] &&
    !targetInstance.constructor.prototype[fieldName]
  ) {
    throw new SchemaFieldError(
      targetInstance.constructor,
      fieldName,
      `Every root schema field must regular class function`
    )
  }
}

function requireSchemaRoot(target: Function, fieldName: string) {
  if (schemaRootsRegistry.has(target)) {
    return
  }
  throw new SchemaFieldError(
    target,
    fieldName,
    `Root field must be registered on class decorated with @Schema`
  )
}

function getFieldCompiler(target: Function, fieldName: string) {
  const fieldCompiler = () => {
    requireSchemaRoot(target, fieldName)
    const compiledField = compileFieldConfig(target, fieldName)
    return compiledField
  }

  return fieldCompiler
}

export enum rootFieldTypes {
  query = 'query',
  mutation = 'mutation'
}

// special fields
export function Query(options?: IFieldOptions): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    validateRootSchemaField(targetInstance, fieldName)
    Field({ rootFieldType: rootFieldTypes.query, ...options })(
      targetInstance,
      fieldName
    )
    const fieldCompiler = getFieldCompiler(
      targetInstance.constructor,
      fieldName
    )
    queryFieldsRegistry.set(
      targetInstance.constructor,
      fieldName,
      fieldCompiler
    )
  }
}

export function Mutation(options?: IFieldOptions): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    validateRootSchemaField(targetInstance, fieldName)
    Field({ rootFieldType: rootFieldTypes.mutation, ...options })(
      targetInstance,
      fieldName
    )
    const fieldCompiler = getFieldCompiler(
      targetInstance.constructor,
      fieldName
    )
    mutationFieldsRegistry.set(
      targetInstance.constructor,
      fieldName,
      fieldCompiler
    )
  }
}

export function QueryAndMutation(options?: IFieldOptions): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    validateRootSchemaField(targetInstance, fieldName)
    Field({ rootFieldType: rootFieldTypes.query, ...options })(
      targetInstance,
      fieldName
    )
    const fieldCompiler = getFieldCompiler(
      targetInstance.constructor,
      fieldName
    )

    queryFieldsRegistry.set(
      targetInstance.constructor,
      fieldName,
      fieldCompiler
    )

    mutationFieldsRegistry.set(
      targetInstance.constructor,
      fieldName,
      fieldCompiler
    )
  }
}
