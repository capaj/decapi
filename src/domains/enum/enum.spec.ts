import { resolveType } from '../../services/utils/gql/types/typeResolvers'
import { registerEnum } from '../..'
import { SchemaRoot, compileSchema } from '../schema/SchemaRoot'
import { Query } from '../schema/rootFields'
import { printSchema, graphql } from 'graphql'
import { Arg } from '../arg/ArgDecorators'

describe('Enums', () => {
  it('Registers returns proper enum type', () => {
    enum Foo {
      'Bar',
      'Baz'
    }

    const enumType = registerEnum(Foo, 'Foo')
    expect(enumType.name).toEqual('Foo')
    expect(enumType.getValues().length).toEqual(2)
    expect(enumType.getValues()[0].name).toEqual('Bar')
    expect(enumType.getValues()[0].value).toEqual(0)
  })

  it('Registers returns proper enum type with string based enums', () => {
    enum Foo {
      Bar = 'Test',
      Baz = 'Test2'
    }

    const enumType = registerEnum(Foo, 'Foo')
    expect(enumType.name).toEqual('Foo')
    expect(enumType.getValues().length).toEqual(2)
    expect(enumType.getValues()[1].name).toEqual('Baz')
    expect(enumType.getValues()[1].value).toEqual('Test2')
  })

  it('Throw when registering the same enum twice', () => {
    enum Foo {
      Bar = 'Test',
      Baz = 'Test2'
    }

    registerEnum(Foo, { name: 'Foo' })
    expect(() =>
      registerEnum(Foo, { name: 'Foo2' })
    ).toThrowErrorMatchingInlineSnapshot(
      `"Enum Foo2: Enum is already registered"`
    )
  })

  it('Will properly resolve registered enum', () => {
    enum Foo {
      Bar = 'Test',
      Baz = 'Test2'
    }

    const enumType = registerEnum(Foo, { name: 'Foo' })
    expect(resolveType(Foo)).toEqual(enumType)
  })

  it('renders schema with an enum used in a query', async () => {
    enum IntEnum {
      one,
      two
    }

    enum StateEnum {
      Done = 'DONE',
      InProgress = 'INPROGRESS',
      Finished = 'FINISHED',
      Cancelled = 'CANCELLED'
    }

    registerEnum(StateEnum, { name: 'StateEnum' })
    registerEnum(IntEnum, { name: 'IntEnum' })
    @SchemaRoot()
    class FooSchema {
      @Query()
      echoAsInferred(@Arg({ type: StateEnum }) input: StateEnum): StateEnum {
        return input
      }
      @Query({ type: StateEnum })
      echoAsEnum(@Arg({ type: StateEnum }) input: StateEnum): StateEnum {
        return input
      }

      @Query()
      intAsInferred(@Arg({ type: IntEnum }) input: IntEnum): IntEnum {
        return input
      }

      @Query({ type: IntEnum })
      intAsEnum(@Arg({ type: IntEnum }) input: IntEnum): IntEnum {
        return input
      }
    }
    const schema = compileSchema(FooSchema)
    expect(printSchema(schema)).toMatchSnapshot()
    const result = await graphql(
      schema,
      `
        {
          echoAsInferred(input: InProgress)
          echoAsEnum(input: InProgress)
          intAsEnum(input: two)
          intAsInferred(input: two)
        }
      `
    )

    expect(result).toMatchInlineSnapshot(`
Object {
  "data": Object {
    "echoAsEnum": "InProgress",
    "echoAsInferred": "INPROGRESS",
    "intAsEnum": "two",
    "intAsInferred": 1,
  },
}
`)
  })
})
