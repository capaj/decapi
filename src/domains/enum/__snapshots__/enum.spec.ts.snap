// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enums renders schema with an enum used in a query 1`] = `
"type Query {
  echoAsInferred(input: StateEnum): String
  echoAsEnum(input: StateEnum): StateEnum
  echoAsEnum2(input: StateEnumUsingKeys): StateEnumUsingKeys
  intAsInferred(input: IntEnum): Float
  intAsEnum(input: IntEnum): IntEnum
}

enum StateEnum {
  DONE
  IN_PROGRESS
  FINISHED
  CANCELLED
}

enum StateEnumUsingKeys {
  Done
  InProgress
  Finished
  Cancelled
}

enum IntEnum {
  one
  two
}"
`;
